File: app/src/main/java/com/softwareoverflow/hiitultimate/viewModel/WorkoutViewModelFactory.java
package com.softwareoverflow.hiitultimate.viewModel;

import android.app.Application;

import androidx.annotation.NonNull;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;

public class WorkoutViewModelFactory implements ViewModelProvider.Factory {

    private Application application;
    private Integer workoutId;

    public WorkoutViewModelFactory(@NonNull Application application, Integer workoutId) {
        this.application = application;
        this.workoutId = workoutId;
    }

    @SuppressWarnings("unchecked")
    @NonNull
    @Override
    public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
        return (T) new WorkoutViewModel(application, workoutId);
    }
}

File: app/src/main/java/com/softwareoverflow/hiit_trainer/ui/workout/WorkoutViewModelFactory.kt
package com.softwareoverflow.hiit_trainer.ui.workout

import android.content.Context
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.preference.PreferenceManager
import com.softwareoverflow.hiit_trainer.repository.dto.WorkoutDTO
import com.softwareoverflow.hiit_trainer.ui.getWorkoutCompleteExerciseType
import com.softwareoverflow.hiit_trainer.ui.workout.media.WorkoutMediaManager

class WorkoutViewModelFactory(
    private val context: Context,
    private val workout: WorkoutDTO
) :
    ViewModelProvider.Factory {

    @Suppress("UNCHECKED_CAST")
    override fun <T : ViewModel> create(modelClass: Class<T>): T {
        if (modelClass.isAssignableFrom(WorkoutViewModel::class.java)) {
            val prefs = PreferenceManager.getDefaultSharedPreferences(context.applicationContext)
            val mediaManager = WorkoutMediaManager(context, prefs)

            return WorkoutViewModel(
                context.applicationContext,
                workout,
                getWorkoutCompleteExerciseType(context.applicationContext),
                mediaManager
            ) as T
        }

        throw IllegalArgumentException("Unknown ViewModel class")
    }
}
